### PR TITLE
Read only user integration tests

### LINKED ISSUE #736: Test client works for readonly users

### PR DESCRIPTION
**Pull request by jkaflik**
Draft a proposal to make it possible to establish a read-only user connection in tests. It can be used to cover further use cases.

Connection initialization may require additional attention in the future if it grows more.

Resolves #736

### CODE REVIEW

File: tests/conn_test.go [hunk 1]
```diff
@@ -242,3 +242,33 @@ func TestBlockBufferSize(t *testing.T) {
 	}
 	require.Equal(t, 10000000, i)
 }
+
+func TestReadOnlyUserConnInsert(t *testing.T) {
```
Comment (mshustov) [inline tests/conn_test.go:246]: We can add other testing scenarios for read-only users:
- should select data from a table
- shouldn't select data from a table 
- shouldn't create a table
- shouldn't delete a table

see https://github.com/ClickHouse/clickhouse-js/blob/01eab4c819992d9f69d94d0d38b952723d1f203c/__tests__/integration/read_only_user.test.ts

File: tests/utils.go [hunk 2]
```diff
@@ -273,7 +281,7 @@ func GetConnectionWithOptions(options *clickhouse.Options) (driver.Conn, error)
 	return clickhouse.Open(options)
 }
 
-func getConnection(env ClickHouseTestEnvironment, database string, settings clickhouse.Settings, tlsConfig *tls.Config, compression *clickhouse.Compression) (driver.Conn, error) {
+func getConnection(env ClickHouseTestEnvironment, database string, roUser bool, settings clickhouse.Settings, tlsConfig *tls.Config, compression *clickhouse.Compression) (driver.Conn, error) {
```
Comment (mshustov) [inline tests/utils.go:284]: Nit: we need something more powerful than `bool`. There are 3 modes for `readonly` users https://clickhouse.com/docs/en/operations/settings/permissions-for-queries/#readonly

File: tests/resources/admin.xml [hunk 3]
```diff
@@ -6,11 +6,20 @@
...
+            <readonly>1</readonly>
+        </readonly>
     </profiles>
     <users>
         <default>
             <password>ClickHouse</password>
             <access_management>1</access_management>
         </default>
+        <readonly>
+            <password>ClickHouse</password>
```
[the remark this answers is not in the document]
Comment (jkaflik) [reply tests/resources/admin.xml:21]: Yes, totally make sense. Will have a look on that.

File: tests/conn_test.go (hunk 1 above)
Comment (mshustov) [reply tests/conn_test.go:246]: @gingerwizard do you have other ideas?

Comment (jkaflik) [reply tests/conn_test.go]: I've added mentioned scenarios with one-two extra.

File: tests/utils.go (hunk 2 above)
Comment (jkaflik) [reply tests/utils.go]: @mshustov definitely. However, I would start with having a read-only user covered. Please take a look at the recent changes.

File: tests/utils.go [hunk 4]
```diff
@@ -245,6 +245,61 @@ func GetExternalTestEnvironment(testSet string) (ClickHouseTestEnvironment, erro
...
+	settings := clickhouse.Settings{}
+
+	if proto.CheckMinVersion(proto.Version{
+		Major: 22,
+		Minor: 8,
+		Patch: 0,
+	}, env.Version) {
+		settings["database_replicated_enforce_synchronous_settings"] = "1"
+	}
+	settings["insert_quorum"], _ = strconv.Atoi(GetEnv("CLICKHOUSE_QUORUM_INSERT", "1"))
```
Comment (mshustov) [inline tests/utils.go:296]: Nit: can we extract env var parsing to a function to avoid repeating this logic?

File: tests/read_only_user_test.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+				assert.Equal(t, expectedValue, uint64(5))
+			},
+		},
+	}
+
+	writeQueryCases := []struct {
+		name             string
+		query            string
+		supportedVersion struct{ major, minor, patch uint64 }
```
Comment (mshustov) [inline tests/read_only_user_test.go:51]: Can it be removed?

File: tests/read_only_user_test.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestReadOnlyUser(t *testing.T) {
+	readQueryCases := []struct {
+		name        string
+		query       string
+		assertRowFn func(row driver.Row)
```
Comment (mshustov) [inline tests/read_only_user_test.go:33]: Could you add a test for users who can't select from a table without access granted?

**Review by mshustov — APPROVED** [review]

File: tests/read_only_user_test.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	}{
+		{
+			name:  "select from table",
+			query: "SELECT sum(Col1) FROM test_readonly_user",
```
[the remark this answers is not in the document]
Comment (jkaflik) [reply tests/read_only_user_test.go:37]: I think asserting the exact value is not a subject of this test, so I would leave it there. We might even assume the table is empty as we want to check only if we have permission to query.

File: tests/read_only_user_test.go (hunk 6 above)
Comment (jkaflik) [reply tests/read_only_user_test.go:33]: I will add a dedicated test case for that in the future.

File: tests/utils.go (hunk 4 above)
Comment (jkaflik) [reply tests/utils.go:296]: I want to refactor `utils.go` in close future.

**Pull request merged**